modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/GridEncryptionManager.java

```diff
@@ -194,10 +207,22 @@
      * Master key change prepare process. Checks that all server nodes have the same new master key and then starts
      * finish process.
      */
-    private DistributedProcess<MasterKeyChangeRequest, MasterKeyChangeResult> prepareMKChangeProc;
+    private DistributedProcess<MasterKeyChangeRequest, EmptyResult> prepareMKChangeProc;
 
     /** Process to perform the master key change. Changes master key and reencrypt group keys. */
-    private DistributedProcess<MasterKeyChangeRequest, MasterKeyChangeResult> performMKChangeProc;
+    private DistributedProcess<MasterKeyChangeRequest, EmptyResult> performMKChangeProc;
+
+    /** Two phase distributed process, that performs cache group encryption key rotation. */
+    private GroupKeyChangeProcess grpKeyChangeProc;
+
+    /** Cache groups for which encryption key was changed, and they must be re-encrypted. */
+    private final Map<Integer, long[]> reencryptGroups = new ConcurrentHashMap<>();
+
+    /** Cache groups for which encryption key was changed on node join. */
+    private final Map<Integer, Integer> reencryptGroupsForced = new ConcurrentHashMap<>();
```

Comment:
Why do we only add entries to this map and never clean it up?

Reply:
fixed